hedyorg/hedy
# [UI] Make keyword toggle also switcher language of the editor

Fixes #4420 

**Some background**

I want to describe a bit of background here because this is such a lovely example of a set of very reasonable decisions leading to bad design!

When we introduced the keyword languages, we added a switcher on the top of the screen, to switch all code examples in the adventures. So far so good. But we also switched start codes in the editor and translated them (since they have {}, the translation does not mean *parsing* but means grabbing from the yaml file and replacing the keywords). This happens when you have not written code yet. If you do, it does not load the startcode and then it does not translate.

But... Since we seems to translate code in the editor... people thought it could also translate their own code. It actually took me a long time to figure out why people (like @Mark-Giesen) thought that that is what the button would do! 

This PR makes the one button translate in the editor too. 

**Limitations**

The error handling in the translation code is currently quite limited. I tried to expand it in https://github.com/hedyorg/hedy/pull/4828/commits/e08375dbcb463d2759e437547fda26d71ddc6470 but it will involve extensive refactoring to do it nicely, so I prefer to do that separately.

I was also planning to Rename the top switcher so that it does not show the translation emoji, but the current keyword language (so: en or nl), but this leads to some design issues (https://github.com/hedyorg/hedy/pull/4828#issuecomment-1833151233) that I also want to address separately.

**Changes**

* Makes the top switcher translate the program too
* Removes the switcher in the editor for the admin 
* Gives an error message in the regular place instead of a shortlived modal
* Show the top switcher, also when people are logged in, so they can use the button to switch it -> was already the case!

**How to test**
Set Hedy to any non-En language and write code in that language (f.e. slaap 20 in Dutch, level 4). Use the toggle and observe the keywords being translated.
Create a broken program with a parse error and see the error message appear below:

![Uploading image.png…]()
